### PR TITLE
Data Import Improvements

### PR DESCRIPTION
**Pull request by SafarMirek**
This PR aims to improve data import into existing registry. It allows users to define headers to disable globalId and contentId persistency of imported data to avoid id conflicts with existing schemas in the registry.

 #1922

### CODE REVIEW

File: common/src/main/java/io/apicurio/registry/rest/v2/beans/SystemInfo.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
     private String description;
     @JsonProperty("version")
     private String version;
-    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss'Z'", timezone = "UTC")
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssZ", timezone = "UTC")
```
Comment (famartinrh) [inline common/src/main/java/io/apicurio/registry/rest/v2/beans/SystemInfo.java:35]: you forgot to revert this change

Comment (SafarMirek) [reply common/src/main/java/io/apicurio/registry/rest/v2/beans/SystemInfo.java:35]: Oh, but, is there a reason why everywhere else is yyyy-MM-dd'T'HH:mm:ssZ and only on this place there is yyyy-MM-dd'T'HH:mm:ss'Z'?

**Review by famartinrh — APPROVED** [review]

File: app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java [hunk 2]
```diff
@@ -2294,16 +2311,67 @@ public void exportData(Function<Entity, Void> handler) throws RegistryStorageExc
...
+        handles.withHandleNoException(handle -> {
             Entity entity = null;
-            while ( (entity = entities.nextEntity()) != null ) {
-                if (entity != null) {
-                    importEntity(handle, entity);
+            while ((entity = entities.nextEntity()) != null) {
+                if (entity.getEntityType() == EntityType.ArtifactVersion) {
+                    ArtifactVersionEntity artifactVersionEntity = (ArtifactVersionEntity) entity;
+                    if (!preserveGlobalId) {
+                        ((ArtifactVersionEntity) entity).globalId = -1;
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:2329]: Why cast again?  Why not just use `artifactVersionEntity`?

File: app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java [hunk 3]
```diff
@@ -2294,16 +2311,67 @@ public void exportData(Function<Entity, Void> handler) throws RegistryStorageExc
...
+                if (entity.getEntityType() == EntityType.Content) {
+                    ContentEntity contentEntity = (ContentEntity) entity;
+                    if (!preserveContentId) {
+                        // When we do not want to preserve contentId, the best solution to import content is create new one with the contentBytes
+                        // It makes sure there won't be any conflicts
+                        long newContentId = -1;
+                        if (contentEntity.artifactType != null) {
+                            newContentId = createOrUpdateContent(handle, contentEntity.artifactType, ContentHandle.create(contentEntity.contentBytes));
+                        } else {
+                            if(contentEntity.canonicalHash == null) throw new RegistryStorageException("There is not enough information about content. Artifact Type and CanonicalHash are both missing.");
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:2350]: Can you format this a bit differently?  Just a style issue:

```java
if (contentEntity.canonicalHash == null) {
    throw new RegistryStorageException("There is not enough information about content. Artifact Type and CanonicalHash are both missing.");
}
```

Just for code style consistency.

File: app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java [hunk 4]
```diff
@@ -2294,16 +2311,67 @@ public void exportData(Function<Entity, Void> handler) throws RegistryStorageExc
...
+        // To handle issue, when we do not know new content id before importing the actual content
+        ArrayList<ArtifactVersionEntity> waitingForContent = new ArrayList<>();
+
+        handles.withHandleNoException(handle -> {
             Entity entity = null;
-            while ( (entity = entities.nextEntity()) != null ) {
-                if (entity != null) {
-                    importEntity(handle, entity);
+            while ((entity = entities.nextEntity()) != null) {
+                if (entity.getEntityType() == EntityType.ArtifactVersion) {
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:2326]: I recommend putting the `ArtifactVersion` logic below the `Content` logic.  It should make the code easier to read/interpret in the future.  Logically we would expect to process the Content entities before the artifact version entities.  But even if we don't (in practice), which should **think** about it like we do.

File: common/src/main/resources/META-INF/openapi.json [hunk 5]
```diff
@@ -720,6 +720,25 @@
...
+                    },
+                    {
+                        "name": "X-Registry-Preserve-ContentId",
+                        "description": "If this header is set to false, content ids of imported data will be ignored and replaced by next id in content id sequence. The mapping between content and artifacts will be preserved. This allows to import any data even thought the content ids would cause a conflict.",
+                        "schema": {
+                            "type": "boolean"
+                        },
+                        "in": "header"
+                    }
+                ],
```
Comment (EricWittmann) [inline common/src/main/resources/META-INF/openapi.json:741]: Have these been added to the API spec in Apicurio Studio?  If yes, which one? :)

File: storage/kafkasql/src/main/java/io/apicurio/registry/storage/impl/kafkasql/KafkaSqlRegistryStorage.java [hunk 6]
```diff
@@ -940,15 +965,66 @@ public void exportData(Function<Entity, Void> handler) throws RegistryStorageExc
...
+                    ContentEntity contentEntity = (ContentEntity) entity;
+                    if (!preserveContentId) {
+                        // When we do not want to preserve contentId, the best solution to import content is create new one with the contentBytes
+                        // It makes sure there won't be any conflicts
+                        long newContentId;
+                        if (contentEntity.artifactType != null) {
+                            newContentId = ensureContentAndGetContentId(ContentHandle.create(contentEntity.contentBytes),  contentEntity.artifactType);
+                        } else {
+                            if (contentEntity.canonicalHash == null)
+                                throw new RegistryStorageException("There is not enough information about content. Artifact Type and CanonicalHash are both missing.");
```
Comment (EricWittmann) [inline storage/kafkasql/src/main/java/io/apicurio/registry/storage/impl/kafkasql/KafkaSqlRegistryStorage.java:1003]: Can you throw some braces around this?  Style issue. :)

**Review by EricWittmann — CHANGES_REQUESTED** [review]
Approved after some minor changes.

**Review by EricWittmann — APPROVED** [review]

**Comment by EricWittmann** [comment]
Nice, thanks!
